File: C/MaximumTwinSumOfLinkedList.c
struct ListNode* reverse(struct ListNode* head) {
    struct ListNode* prev = NULL;
    
    while (head) {
        struct ListNode* next = head->next;
        head->next = prev;
        prev = head;
        head = next;
    }
    
    return prev;
}

int pairSum(struct ListNode* head) {
    struct ListNode *slow = head, *fast = head;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    struct ListNode* second = reverse(slow);
    int ans = 0;

    while (second) {
        int sum = head->val + second->val;
        if (sum > ans) ans = sum;

        head = head->next;
        second = second->next;
    }

    return ans;
}

File: Python/MaximumTwinSumOfLinkedlist.py
class Solution(object):
    def pairSum(self, head):
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        second_half = self.reverse(slow)

        ans = 0
        first_half = head

        while second_half:
            ans = max(ans, first_half.val + second_half.val)
            first_half = first_half.next
            second_half = second_half.next

        return ans

    def reverse(self, head):
        prev = None
        curr = head

        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        return prev